compiler/pipes/code-gen.cpp

```diff
@@ -65,6 +66,10 @@ void CodeGenF::on_finish(DataStream<std::unique_ptr<CodeGenRootCmd>> &os) {
       continue;
     }
 
+    if (!c->json_encoders.empty()) {
+      code_gen_start_root_task(os, std::make_unique<JsonEncoderTags>(c->json_encoders));
```

Comment:
If one and the same encoder is used to encode N multiple classes (which is a common case), this line will launch N codegen tasks for one and the same encoder, which is wrong.
I'd suggest to collect all encoders into an `std::set<ClassPtr> all_json_encoders` or similar (named like `all_classes` above). And to launch `JsonEncoderTags` once afterward, providing that set.

Reply:
fixed